### PR TITLE
Refactor cpp generator

### PR DESCRIPTION
**Pull request by sigiesec**
This change is [Reviewable](https://reviewable.io/reviews/btc-ag/service-idl/11)

### CODE REVIEW

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+import static extension com.btc.serviceidl.generator.cpp.Util.*
+import static extension com.btc.serviceidl.util.Extensions.*
+import java.util.Comparator
+
+@Accessors(NONE)
+class BasicCppGenerator
+{
+    // must be protected to allow subclasses to benefit from extension declaration
+    @Accessors(PUBLIC_GETTER) protected val extension TypeResolver typeResolver
+    @Accessors(PUBLIC_GETTER) private val ParameterBundle paramBundle
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend]: Visibility is private by default in Xtend, may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    def dispatch String toText(PrimitiveType item, EObject context)
+    {
+        getPrimitiveTypeName(item)
+    }
+
+    def dispatch String toText(SequenceDeclaration item, EObject context)
+    {
+        val inner_type = '''«IF item.failable»«
+
+resolveSymbol
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend]: Lines 251-254: indentation level seems to be wrong.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/CppGenerator.xtend [hunk 3]
```diff
@@ -1,3836 +1,178 @@
...
-         );
-         
-         template<typename API_TYPE, typename PROTOBUF_TYPE>
-         void EncodeFailable
-         (
-            «forward_const_iterator»< «failable_handle»<API_TYPE> > api_input,
-            google::protobuf::RepeatedPtrField< PROTOBUF_TYPE >* const protobuf_output
-         );
-         
-         template<typename PROTOBUF_TYPE, ty
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/CppGenerator.xtend]: Line 41 and below: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ProjectGeneratorBaseBase.xtend [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import org.eclipse.xtext.naming.IQualifiedNameProvider
+import org.eclipse.xtext.scoping.IScopeProvider
+
+import static extension com.btc.serviceidl.generator.cpp.CppExtensions.*
+import static extension com.btc.serviceidl.util.Util.*
+
+@Accessors(PROTECTED_GETTER)
+class ProjectGeneratorBaseBase
+{
+    private val Resource resource
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ProjectGeneratorBaseBase.xtend]: This line and below: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/TypeResolver.xtend [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+import org.eclipse.xtext.naming.IQualifiedNameProvider
+
+import static extension com.btc.serviceidl.generator.common.Extensions.*
+import static extension com.btc.serviceidl.generator.cpp.CppExtensions.*
+import static extension com.btc.serviceidl.util.Util.*
+
+@Accessors(NONE)
+class TypeResolver
+{
+    @Accessors(PACKAGE_GETTER) private val IQualifiedNameProvider qualified_name_provider
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/TypeResolver.xtend]: Here and below: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/OdbGenerator.xtend [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+import org.eclipse.xtend.lib.annotations.Accessors
+
+import static extension com.btc.serviceidl.generator.cpp.Util.*
+import static extension com.btc.serviceidl.util.Extensions.*
+import static extension com.btc.serviceidl.util.Util.*
+
+@Accessors
+class OdbGenerator
+{
+    private val extension TypeResolver typeResolver
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/OdbGenerator.xtend]: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/PrinsHeaderResolver.xtend [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ **********************************************************************/
+package com.btc.serviceidl.generator.cpp.prins
+
+import com.btc.serviceidl.generator.cpp.HeaderResolver
+import com.btc.serviceidl.generator.cpp.TypeResolver
+import java.util.Arrays
+
+class PrinsHeaderResolver
+{
+    private static val ODB_INCLUDE_GROUP = new TypeResolver.IncludeGroup("ODB")
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/PrinsHeaderResolver.xtend]: Here and below: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/ReferenceResolver.xtend [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import com.btc.serviceidl.generator.cpp.HeaderResolver
+import com.btc.serviceidl.generator.cpp.TypeResolver
+import org.eclipse.core.runtime.IPath
+
+import static extension com.btc.serviceidl.generator.common.GeneratorUtil.*
+
+class ReferenceResolver
+{
+    // ******************************* PLEASE ALWAYS KEEP THIS LIST ALPHABETICALLY SORTED !!! ******************************* //
+    private static val vs_projects_mapper = #{
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/ReferenceResolver.xtend]: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VSSolution.xtend [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.UUID
+import org.eclipse.core.runtime.IPath
+import org.eclipse.xtend.lib.annotations.Data
+
+class VSSolution implements IProjectSet
+{
+// it is important for this container to be static! if an *.IDL file contains
+// "import" references to external *.IDL files, each file will be generated separately
+// but we need consistent project GUIDs in order to create valid project references!
+    private static val vs_projects = new HashMap<String, Entry>
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VSSolution.xtend]: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VcxProjGenerator.xtend [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+import java.util.Set
+import org.eclipse.core.runtime.IPath
+import org.eclipse.xtend.lib.annotations.Accessors
+
+import static extension com.btc.serviceidl.generator.common.FileTypeExtensions.*
+
+@Accessors
+class VcxProjGenerator
+{
+    private val ParameterBundle param_bundle
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VcxProjGenerator.xtend]: Here and below: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/AssemblyInfoGenerator.xtend [hunk 11]
```diff
@@ -12,7 +12,7 @@ import static extension com.btc.serviceidl.generator.common.Extensions.*
 @Accessors
 class AssemblyInfoGenerator
 {
-    private val ParameterBundle.Builder param_bundle
+    private val ParameterBundle param_bundle
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/AssemblyInfoGenerator.xtend]: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/VSSolution.xtend [hunk 12]
```diff
@@ -5,14 +5,15 @@ import com.btc.serviceidl.generator.common.GeneratorUtil
 import com.btc.serviceidl.generator.common.TransformType
 import java.util.UUID
 import java.util.HashMap
+import com.btc.serviceidl.generator.common.ArtifactNature
 
 class VSSolution
 {
     private val vs_projects = new HashMap<String, UUID>
```
Comment (huttenlocher) [inline com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/VSSolution.xtend]: Visibility is private by default in Xtend, "private" modifier may be omitted.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend (hunk 1 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend]: Removed (on all occurrences)

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend (hunk 2 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/BasicCppGenerator.xtend]: Fixed

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/CppGenerator.xtend (hunk 3 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/CppGenerator.xtend]: The comment appears to be misplaced here, but I removed the private declaration in all files.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ProjectGeneratorBaseBase.xtend (hunk 4 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/ProjectGeneratorBaseBase.xtend]: Removed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/TypeResolver.xtend (hunk 5 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/TypeResolver.xtend]: Removed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/OdbGenerator.xtend (hunk 6 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/OdbGenerator.xtend]: Removed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/OdbGenerator.xtend [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    def generateODBTraitsBody()
+    {
+        '''
+            namespace odb
+            {
+               // ***** MSSQL *****
+               namespace mssql
+               {
+                  template<>
+                  struct default_type_traits<«resolveSymbol("BTC::PRINS::Commons::GUID")»>
```
[the remark this answers is not in the document]
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/OdbGenerator.xtend]: I didn't want to make any (intentional) changes to the generated code at this moment. I added a comment.

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/PrinsHeaderResolver.xtend (hunk 7 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/PrinsHeaderResolver.xtend]: Fixed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/ReferenceResolver.xtend (hunk 8 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/ReferenceResolver.xtend]: Fixed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VSSolution.xtend (hunk 9 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VSSolution.xtend]: Fixed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VcxProjGenerator.xtend (hunk 10 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/cpp/prins/VcxProjGenerator.xtend]: Fixed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/AssemblyInfoGenerator.xtend (hunk 11 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/AssemblyInfoGenerator.xtend]: Fixed everywhere

File: com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/VSSolution.xtend (hunk 12 above)
Comment (sigiesec) [reply com.btc.serviceidl/src/com/btc/serviceidl/generator/dotnet/VSSolution.xtend]: Fixed everywhere

**Comment by sigiesec** [comment]
@huttenlocher I am not what you looked at. The whole PR was huge and the commits splitting up the huge CppGenerator class as well, but the remaining commits were much smaller each.

Follow-up PRs will be smaller overall, hopefully :)

**Pull request merged**